modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItIdempotentCommandCacheTest.java

```diff
@@ -0,0 +1,224 @@
+
+        storage = new RocksDbKeyValueStorage(
+                clusterService.nodeName(),
+                workDir.resolve("metastorage"),
+                new NoOpFailureProcessor(clusterService.nodeName()));
+
+        metaStorageManager = new MetaStorageManagerImpl(
+                clusterService,
+                cmgManager,
+                logicalTopologyService,
+                raftManager,
+                storage,
+                clock,
+                topologyAwareRaftGroupServiceFactory,
+                metaStorageConfiguration
+        );
+
+        assertThat(
+                startAsync(clusterService, raftManager, metaStorageManager)
+                        .thenCompose(unused -> metaStorageManager.deployWatches()),
+                willCompleteSuccessfully()
+        );
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        List<IgniteComponent> components = List.of(metaStorageManager, raftManager, clusterService);
+
+        closeAll(Stream.concat(
+                components.stream().map(c -> c::beforeNodeStop),
+                Stream.of(() -> assertThat(stopAsync(components), willCompleteSuccessfully()))
+        ));
+    }
+
+    @Test
+    public void testIdempotentInvoke() throws InterruptedException {
+        ByteArray testKey = new ByteArray("key".getBytes(StandardCharsets.UTF_8));
+        byte[] testValue = "value".getBytes(StandardCharsets.UTF_8);
+
+        AtomicInteger writeActionReqCount = new AtomicInteger();
+        CompletableFuture<Void> retryBlockingFuture = new CompletableFuture<>();
+
+        log.info("Test: blocking messages.");
+
+        dropMessages((n, msg) -> {
+            // Dropping the first response, this will cause timeout on first response, and then retry.
+            if (msg instanceof ActionResponse && ((ActionResponse) msg).result() != null && writeActionReqCount.get() == 1) {
+                log.info("Test: dropping ActionResponse: " + msg);
+
+                return true;
+            }
+
+            if (msg instanceof WriteActionRequest) {
+                WriteActionRequest request = (WriteActionRequest) msg;
+
+                if (!(request.deserializedCommand() instanceof SyncTimeCommand)) {
+                    writeActionReqCount.incrementAndGet();
+                    log.info("Test: WriteActionRequest intercepted, count=" + writeActionReqCount.get());
+
+                    // Second request: retry.
+                    if (writeActionReqCount.get() == 2) {
+                        log.info("Test: retry blocked.");
+
+                        retryBlockingFuture.orTimeout(10, TimeUnit.SECONDS).join();
+
+                        log.info("Test: retry unblocked.");
+                    }
+                }
+            }
+
+            return false;
+        });
+
+        CompletableFuture<Boolean> fut = metaStorageManager.invoke(notExists(testKey), put(testKey, testValue), noop());
```

Comment:
I'd also have put(anotherTestKey,...) instead of noop() and corresponding check that it won't be actually putted.

Reply:
Done